src/transport/ExchangeMgr.h

```diff
@@ -0,0 +1,270 @@
+    uint32_t RetransInterval;      /**< Time between retransmissions (in milliseconds); 0 disables retransmissions. */
+    Timeout ResponseTimeout;       /**< Maximum time to wait for response (in milliseconds); 0 disables response timeout. */
+
+    enum
+    {
+        kSendFlag_AutoRetrans    = 0x0001, /**< Used to indicate that automatic retransmission is enabled. */
+        kSendFlag_ExpectResponse = 0x0002, /**< Used to indicate that a response is expected within a specified timeout. */
+        kSendFlag_DelaySend      = 0x0008, /**< Used to indicate that the sending of the current message needs to be delayed. */
+        kSendFlag_ReuseSourceId  = 0x0020, /**< Used to indicate that the source node ID in the message header can be reused. */
+        kSendFlag_RetainBuffer   = 0x0040, /**< Used to indicate that the message buffer should not be freed after sending. */
+        kSendFlag_AlreadyEncoded = 0x0080, /**< Used to indicate that the message is already encoded. */
+        kSendFlag_DefaultMulticastSourceAddress = 0x0100, /**< Used to indicate that default IPv6 source address selection should be
+                                                             used when sending IPv6 multicast messages. */
+        kSendFlag_FromInitiator    = 0x0200, /**< Used to indicate that the current message is the initiator of the exchange. */
+        kSendFlag_RequestAck       = 0x0400, /**< Used to send a RMP message requesting an acknowledgment. */
+        kSendFlag_NoAutoRequestAck = 0x0800, /**< Suppress the auto-request acknowledgment feature when sending a message. */
+
+        kSendFlag_MulticastFromLinkLocal = kSendFlag_DefaultMulticastSourceAddress,
+        /**< Deprecated alias for \c kSendFlag_DefaultMulticastSourceAddress */
+    };
+
+    bool IsInitiator() const;
+    bool IsResponseExpected() const;
+    void SetInitiator(bool inInitiator);
+    void SetResponseExpected(bool inResponseExpected);
+    bool AutoRequestAck() const;
+    void SetAutoRequestAck(bool autoReqAck);
+
+    CHIP_ERROR SendMessage(uint32_t protocolId, uint8_t msgType, System::PacketBuffer * msgPayload, uint16_t sendFlags = 0,
+                           void * msgCtxt = nullptr);
+    CHIP_ERROR SendCommonNullMessage();
+
+    void TeardownTrickleRetransmit();
+
+    /**
+     * This function is the application callback for handling a received CHIP message.
+     *
+     *  @param[in]    ec            A pointer to the ExchangeContext object.
+     *
+     *  @param[in]    packetHeader  A reference to the PacketHeader object.
+     *
+     *  @param[in]    protocolId    The protocol identifier of the received message.
+     *
+     *  @param[in]    msgType       The message type of the corresponding profile.
+     *
+     *  @param[in]    payload       A pointer to the PacketBuffer object holding the message payload.
+     */
+    typedef void (*MessageReceiveFunct)(ExchangeContext * ec, const PacketHeader & packetHeader, uint32_t protocolId,
```

Comment:
would these look better as pure virtual methods on the application state object?

Reply:
I guess you mean define a Delegate interface for ExchangeContext, right? Currently, all protocols use direct callback. I keep this interface aligned with wml to prevent us from updating all protocols implementation.

Reply:
Maybe we should refactor the other protocols to use delegates. Most other CHIP APIs (Controller, SecureSessionMgr, Rendezvous, Pairing etc) are moving towards using delegates.

Reply:
Done, changed to delegate pattern to align with the rest of code.